Observation (#1998)
apple/swift-evolution: proposals/NNNN-observability.md

        }
        set {
            _$observationRegistrar.withMutation(self, keyPath: \.account) {
                _$observationStorage.account = newValue
            }
        }
    }

    var alternateIconsUnlocked: Bool {
        get { 
            _$observationRegistrar.access(self, keyPath: \.alternateIconsUnlocked)
            return _$observationStorage.alternateIconsUnlocked
        }
        set {
            _$observationRegistrar.withMutation(self, keyPath: \.alternateIconsUnlocked) {
                _$observationStorage.alternateIconsUnlocked = newValue
            }
        }
    }

    var allRecipesUnlocked: Bool {
        get { 
            _$observationRegistrar.access(self, keyPath: \.allRecipesUnlocked)
            return _$observationStorage.allRecipesUnlocked
        }
        set {
            _$observationRegistrar.withMutation(self, keyPath: \.allRecipesUnlocked) {
                _$observationStorage.allRecipesUnlocked = newValue
            }
        }
    }
  
    init() { }
  
    func purchase(alternateIcons: Bool, allRecipes: Bool) {
        alternateIconsUnlocked = alternateIcons
        allRecipesUnlocked = allRecipes
    }
}
```

> When a property does not have a default value that corisponding argument in the initializer does not have a default value. This means that the following example has a macro synthesized initializer of `init(a: Int, b: Int = 3)`.

Reviewer: ```suggestion
When a property does not have a default value that corresponding argument in the initializer does not have a default value. This means that the following example has a macro synthesized initializer of `init(a: Int, b: Int = 3)`.
```

}
```

Properties that have `willSet` and `didSet` property observations will be supported.

```swift
@Observable final class PropertyExample {
    var a: Int {
        willSet { print("will set triggered") }
        didSet { print("did set triggered") }
    }
}
```

Transforms into the macro expansion for the property `a` as follows:

```swift
var a: Int {
    get {
        _$observationRegistrar.access(self, keyPath: \.a)
        return _$observationStorage.a 
    }
    set {
        print("will set triggered")
        _$observationRegistrar.withMutation(of: self, keyPath: \.a) {
            _$observationStorage.a = newValue
        }
        print("did set triggered")
    }
}
```

### `TrackedProperties`

When observing changes to a type, there may be associated side effects to members that are not publicly visible. The `TrackedProperties` type allows for internal key paths to be included in a transaction without being exposed beyond their visibility. 

```swift
public struct TrackedProperties<Root>: ExpressibleByArrayLiteral, @unchecked Sendable {

>     public typealias Element = PartialKeyPath<Root>

Reviewer: This declaration seems unnecessary? It seems to have been added to support SetAlgebra but the protocol conformance isn't there, and the alias isn't used otherwise.

            return _$observationStorage.allRecipesUnlocked
        }
        set {
            _$observationRegistrar.withMutation(self, keyPath: \.allRecipesUnlocked) {
                _$observationStorage.allRecipesUnlocked = newValue
            }
        }
    }
  
    nonisolated static func dependencies(
        of keyPath: PartialKeyPath<Self>
    ) -> TrackedProperties<Self> {
        [keyPath]
    }
    
    func purchase(alternateIcons: Bool, allRecipes: Bool) {
        alternateIconsUnlocked = alternateIcons
        allRecipesUnlocked = allRecipes
    }
}
```

When a property does not have a default value, that corresponding argument in the initializer does not have a default value. This means that the following example has a macro synthesized initializer of `init(a: Int, b: Int = 3)`.

```swift
@Observable final class InitializationSample {
    var a: Int
    var b: Int = 3
}
```

Because the memberwise initializer and backing storage are generated together, the initializer is able to initialize that storage. User-defined initializers should call the generated memberwise initializer instead of attempting to initialize the properties directly.

> Due to limitations with macros, all fields must have type information; this restriction may be able to be lifted later when the type system grows a mechanism to detect inferred types.

Reviewer 2: ```suggestion
Since macros are only syntactic transformations, all fields must have explicit type information. This restriction may be able to be lifted later when the type system grows a mechanism to detect inferred types.
```
